### PR TITLE
Re-implement DoubleByteBuf as a simple holder of a pair of buffers

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

To avoid the overhead in copying the components of a DoubleByteBuf into a temp buffer, we should treat it as a simple holder of a pair of buffers, without trying to mimic the `ByteBuf` interface. 

That makes it for a much simpler implementation with no overhead.

### CODE REVIEW

**Comment by sschepens** [comment]
These chances are great, I was thinking of making similar ones today.
Maybe se could replace `coalesce` with a method only available forma tests, to make ir clear that it should only be used for tests

File: pulsar-broker/src/main/java/org/apache/pulsar/client/api/RawMessage.java [hunk 1]
```diff
@@ -48,7 +49,7 @@
      * Serialize a raw message to a ByteBuf. The caller is responsible for releasing
      * the returned ByteBuf.
```
Comment (sschepens) [inline pulsar-broker/src/main/java/org/apache/pulsar/client/api/RawMessage.java]: These docs should mention DoubleByteBuf

File: pulsar-common/src/main/java/org/apache/pulsar/common/api/DoubleByteBuf.java [hunk 2]
```diff
@@ -16,54 +16,23 @@
...
-import io.netty.util.ResourceLeakDetectorFactory;
-import io.netty.util.ResourceLeakTracker;
+import io.netty.util.ReferenceCounted;
 
 /**
  * ByteBuf that holds 2 buffers. Similar to {@see CompositeByteBuf} but doesn't allocate list to hold them.
  */
-@SuppressWarnings("unchecked")
-public final class DoubleByteBuf extends AbstractReferenceCountedByteBuf {
+public final class DoubleByteBuf extends AbstractReferenceCounted {
```
Comment (sschepens) [inline pulsar-common/src/main/java/org/apache/pulsar/common/api/DoubleByteBuf.java]: I would probably rename this to `ByteBufPair` or something of the sort

File: pulsar-common/src/main/java/org/apache/pulsar/common/api/DoubleByteBuf.java [hunk 3]
```diff
@@ -101,385 +65,50 @@ public ByteBuf getSecond() {
...
-        return result + ", components=2)";
-    }
-
-    @Override
-    public ByteBuffer[] nioBuffers() {
-        return nioBuffers(readerIndex(), readableBytes());
+    /**
+     * @return a single buffer with the content of both individual buffers
+     */
+    public ByteBuf coalesce() {
```
Comment (sschepens) [inline pulsar-common/src/main/java/org/apache/pulsar/common/api/DoubleByteBuf.java]: I would remove this method un favor of a static method for tests

File: pulsar-common/src/main/java/org/apache/pulsar/common/api/DoubleByteBuf.java [hunk 4]
```diff
@@ -101,385 +65,50 @@ public ByteBuf getSecond() {
...
-                break;
-            case ADVANCED:
-            case PARANOID:
-                leak = leakDetector.track(buf);
-                if (leak != null) {
-                    return advancedLeakAwareByteBufConstructor.newInstance(buf, leak);
-                }
-                break;
+    @Sharable
+    public static class Encoder extends ChannelOutboundHandlerAdapter {
```
Comment (sschepens) [inline pulsar-common/src/main/java/org/apache/pulsar/common/api/DoubleByteBuf.java]: :+1:

File: pulsar-common/src/main/java/org/apache/pulsar/common/api/DoubleByteBuf.java (hunk 2 above)
Comment (merlimat) [reply pulsar-common/src/main/java/org/apache/pulsar/common/api/DoubleByteBuf.java]: Yes, sounds better. Will rename

File: pulsar-broker/src/main/java/org/apache/pulsar/client/api/RawMessage.java (hunk 1 above)
Comment (merlimat) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/api/RawMessage.java]: Fixed

File: pulsar-common/src/main/java/org/apache/pulsar/common/api/DoubleByteBuf.java (hunk 3 above)
Comment (merlimat) [reply pulsar-common/src/main/java/org/apache/pulsar/common/api/DoubleByteBuf.java]: Good point. I'll also expand the doc to mention it.

**Comment by sschepens** [comment]
Tests seem to pass, the only thing I'm concerned with is reference counting, we're not retaining the buffers when we get a `ByteBufPair` but we're releasing them when we release the pair, but when writing the bufs they should released as well, I think there would be a double free, but I could be wrong

**Comment by merlimat** [comment]
> the only thing I'm concerned with is reference counting, we're not retaining the buffers when we get a ByteBufPair but we're releasing them when we release the pair,

This is similar to the previous behavior, just a bit simplified. The main goal is that creating a ByteBufPair and writing it to the channel:

```java
ByteBuf b = ... 
// b.refCnt() == 1
ctx.writeAndFlush(b); 
// When write is completed the object is automatically 
// released and  buffer is returned to pool
```

With ByteBufPair it will be the same: 

```java
ByteBuf b1, b2;
// b1.refCnt() == 1
// b2.refCnt() == 1

ByteBufPair p = ByteBufPair.get(b1, b2);
// Still
// b1.refCnt() == 1
// b2.refCnt() == 1
// b.refCnt() == 1

ctx.writeAndFlush(b);

// When writing, b1 and b2 get also incremented before the write 
// to compensate the intrinsic release
```

At the end, `b` gets released and deallocated, which will trigger the release of `b1` and `b2`.

Bottom line, when doing `ByteBufPair.get(b1, b2)` the current ownership (of 1 ref-count) for both `b1` and `b2` is transferred to the `ByteBufPair` instance. This is same behavior as previously with the `DoubleByteBuf`. The only difference is that previously we were doing `retain()` and `realease(2)`, but that doesn't change the counter math.

**Pull request merged**